File: LEETCODE/trees/easy/s1_PreorderIterative.java
package LEETCODE.trees.easy;
import java.util.*;

import LEETCODE.trees.BinaryTree;

// in this we will use iterative not recurssion  
// in this we will use stack   

public class s1_PreorderIterative {
    private static List<Integer> preorder(BinaryTree.Node root){

        // root  left right 
        Stack<BinaryTree.Node> stack = new Stack<>();
        List<Integer> list = new ArrayList<>();
        if(root == null){
            return list;
        }

        stack.push(root);
        while (!stack.isEmpty()) {
            
            BinaryTree.Node curr = stack.pop();
            list.add(curr.data);

            if(curr.right != null){
                stack.push(curr.right);
            }
            if(curr.left != null){
                stack.push(curr.left);
            }
            
        }
        return list;
    }
    public static void main(String[] args) {

        preorder(null);
    }
}

File: LEETCODE/trees/IterativeTraversals/s1_PreorderIterative.cpp
#include<bits/stdc++.h>
using namespace std;

struct Node {
    int val;
    Node* left;
    Node* right;
    Node(int val){
        this->val = val;
        this->left = nullptr;
        this->right = nullptr;
    };
};
vector<int> preorder(Node* root){

    stack<Node*> st;
    st.push(root);
    vector<int> ans;

    while (!st.empty()){
        int size = st.size();
        
        for(int i = 0; i < size; i++){
            auto curr = st.top();
            st.pop();
            ans.push_back(curr->val);
            if(curr->right != nullptr){
                st.push(curr->right);
            }
            if(curr->left != nullptr){
                st.push(curr->right);
            }
        }
    }
    return ans;
}
int main(){
    
}